File: Sudoku.py
from collections import deque

class Sudoku:
    def __init__(self, board):
        self.board = board

    def print_board(self): 
        for row in self.board:
            for num in row:
                if num == 0:
                    print(' ',end=' ')
                else:
                    print(num,end=' ')
            print()
    
    def is_valid(self, row, col, num):
        #檢查行與列
        for i in range(9):
            if self.board[row][i] == num:
                return False
        for i in range(9):
            if self.board[i][col] == num:
                return False

        # 檢查3x3
        start_row = 3 * (row // 3)
        start_col = 3 * (col // 3)
        for i in range(start_row, start_row + 3):
            for j in range(start_col, start_col + 3):
                if self.board[i][j] == num:
                    return False
        return True
        
    def find_empty(self):
        for row in range(9):
            for col in range(9):
                if self.board[row][col] == 0:
                    #直接回傳第一個找到的空白
                    return (row, col)
        #如果找不到空位，返回None
        return None
     
    #回朔法(stack方法)
    def backtrack(self):
        stack = []
        empty = self.find_empty()
        
        #沒有空白，代表已完成
        if not empty:
            return True
        
        row, col = empty
        num = 1
        while True:
            #嘗試填數字
            while num <= 9:
                if self.is_valid(row, col, num):
                    
                    #填上當前數字並把當前狀態推入stack
                    self.board[row][col] = num
                    stack.append((row, col, num))
                    
                    #找下一個空格
                    empty = self.find_empty()
                    if not empty:
                        #完成數獨
                        return True
                    
                    row, col = empty
                    #下一個空格再從1重新開始填
                    num = 1
                    break
                #num數到9，還有空白但已經沒得填 => num += 1 -> 10，回朔回上一部 
                num += 1
            
            #回溯
            if num > 9:
                if not stack:
                    #無解
                    return False
                
                #回溯到上一個狀態
                #從stack取出上一步的狀態
                prev_row, prev_col, prev_num = stack.pop()
                #將上一步恢復為0
                self.board[prev_row][prev_col] = 0
                #更新當前位置為上一步的位置
                row, col = prev_row, prev_col
                #嘗試填入上一步數字的下一個數字(以免又填入相同的數字)
                num = prev_num + 1
    def DFS(self):
        empty = self.find_empty()
        if not empty:
            return True  # 解決成功

        row, col = empty
        for num in range(1, 10):
            if self.is_valid(row, col, num):
                self.board[row][col] = num

                if self.DFS():  # 遞歸
                    return True

                # 回溯
                self.board[row][col] = 0
        
        return False  # 解決失敗
    
board1 = [
        [0, 0, 0, 8, 0, 6, 0, 0, 0],
        [0, 0, 0, 0, 2, 0, 0, 7, 0],
        [0, 0, 9, 4, 0, 0, 5, 0, 0],
        [4, 0, 0, 0, 0, 0, 3, 0, 6],
        [3, 0, 0, 0, 0, 0, 0, 0, 0],
        [2, 6, 0, 0, 0, 1, 7, 0, 0],
        [9, 3, 5, 0, 1, 0, 0, 4, 0],
        [0, 0, 0, 5, 0, 8, 2, 0, 0],
        [0, 0, 0, 0, 0, 3, 0, 1, 0],
    ]

board2 =[
        [0,0,2,0,0,0,0,8,4],
        [0,0,1,6,0,0,0,0,7],
        [5,4,9,8,2,7,0,1,3],
        [0,1,5,0,0,0,3,7,8],
        [7,0,3,0,0,5,4,0,9],
        [9,2,0,0,0,0,5,0,0],
        [1,5,0,4,0,0,0,0,2],
        [0,0,0,7,0,3,0,9,6],
        [0,0,6,0,1,0,0,0,5]
    ]

game = Sudoku(board=board1)
game.print_board()
game.DFS()
game.print_board()

File: Sudoku.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <string>

using namespace std;

class Sudoku
{
private:
    vector<vector<int>> m_board;
public:
    Sudoku(vector<vector<int>> board)
    {
        m_board = board;
    }
    void printBoard()
    {
        string buffer = "+ - + - + - + - + - + - + - + - + - +";
        cout<<buffer<<endl;
        for(int i=0; i<9; i++)
        {
            for(int j=0; j<9; j++)
            {
                cout<<"| "<<m_board[i][j]<<" ";
            }
            cout<<"|"<<endl<<buffer<<endl;
        }
    }
    bool isValid(const int& row,const int& col,const int& num)
    {
        for(int i=0; i<9; i++)
        {
            if(m_board[row][i] == num)
                return false;
        }
        for(int i=0; i<9; i++)
        {
            if(m_board[i][col] == num)
                return false;
        }


        int start_row = 3 * (row / 3);
        int start_col = 3 * (col / 3);
        for(int i=start_row; i<start_row+3; i++)
        {
            for(int j=start_col; j<start_col+3; j++)
            {
                if( m_board[i][j] == num)
                    return false;
            }
        }
        return true;
    }

    pair<int, int> find_empty()
    {
        pair<int, int> pos(9,9);    //回傳(9,9)=>None
        for(int i=0; i<9; i++)
        {
            for(int j=0; j<9; j++)
            {
                if(m_board[i][j] == 0)
                {
                    pos = pair<int,int>(i,j);
                    return pos;
                }
            }
        }
        return pos;
    }
    bool backTracking()
    {
        stack<vector<int>> states;
        pair<int, int> empty = find_empty();
        //沒有空白
        if(empty == pair<int, int>(9,9))
            return true;
        int row = empty.first;
        int col = empty.second;
        int num = 1;
        while(1)
        {
            while(num<=9)
            {
                if(isValid(row,col,num))
                {
                    m_board[row][col] = num;
                    states.push({row,col,num});
                    empty = find_empty();
                    if(empty == pair<int, int>(9,9))
                        return true;
                    row = empty.first;
                    col = empty.second;
                    num = 1;
                    break;
                }
                num++;
            }
            if(num>9)
            {
                if(states.empty())
                {
                    return false;
                }
                vector<int> pre_state = states.top();
                states.pop();
                m_board[pre_state[0]][pre_state[1]] = 0;
                row = pre_state[0];
                col = pre_state[1];
                num = pre_state[2] + 1;
            }
        }
    }
};


int main()
{
    vector<vector<int>> board = {
        {0, 0, 0, 8, 0, 6, 0, 0, 0},
        {0, 0, 0, 0, 2, 0, 0, 7, 0},
        {0, 0, 9, 4, 0, 0, 5, 0, 0},
        {4, 0, 0, 0, 0, 0, 3, 0, 6},
        {3, 0, 0, 0, 0, 0, 0, 0, 0},
        {2, 6, 0, 0, 0, 1, 7, 0, 0},
        {9, 3, 5, 0, 1, 0, 0, 4, 0},
        {0, 0, 0, 5, 0, 8, 2, 0, 0},
        {0, 0, 0, 0, 0, 3, 0, 1, 0},
    };
    Sudoku game(board);
    game.printBoard();
    cout<<"<===================================>"<<endl;
    game.backTracking();
    game.printBoard();
    return 0;
}